### PR TITLE
Ensure the correct Taxonomy subclass is used

### PR DESCRIPTION
**Pull request by pomegranited**
## Description

Ensures that we always use the correct Taxonomy model subclass whenever returning or using Taxonomy instances.

## Supporting information

Proposed way to address https://github.com/openedx/edx-platform/pull/32518#discussion_r1244335301

## Testing instructions

Manual testing is possible but not very useful, since we don't yet have views for these models and APIs.

But you can check that the added tests cover the expected polymorphism behavior supported by this change, and that tests are still covering 100% of the `openedx_tagging` module.

## Other information

Here we add a basic SystemTaxonomy class to help validate this change, but more functionality for system taxonomies will be added by a subsequent PR.

Author to do before merge:

* Address [ObjectTag.taxonomy issue](https://github.com/openedx/openedx-learning/pull/60#discussion_r1246098521)

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: openedx_tagging/core/tagging/models.py [hunk 1]
```diff
@@ -402,7 +434,16 @@ def is_valid(self) -> bool:
...
         """
-        return self.taxonomy_id and self.taxonomy.validate_object_tag(self)
+        if self.taxonomy_id:
+            # self.taxonomy doesn't return the correct subclass, so we must fetch it
+            # FIXME:
+            # * cache to prevent re-fetching the taxonomy
+            # * discourage people from using self.taxonomy
+            taxonomy = (
+                Taxonomy.objects.filter(id=self.taxonomy_id).select_subclasses().first()
+            )
```
Comment (pomegranited) [inline openedx_tagging/core/tagging/models.py:444]: What's the best way to do this?  We can:

* cache the fetched taxonomy subclass using a getter property, and
* rename the `taxonomy` field to `_taxonomy`

Comment (ChrisChV) [reply openedx_tagging/core/tagging/models.py:444]: :+1: for this way

**Comment by pomegranited** [comment]
Closed in favor of https://github.com/openedx/openedx-learning/pull/62

**Comment by openedx-webhooks** [comment]
@pomegranited Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-learning&entry.867055334=https://github.com/openedx/openedx-learning/pull/60&entry.1484655318=https://github.com/pomegranited&entry.752974735=2023-06-29+04:44&entry.1917517419=2023-07-07+05:44&entry.2133058324=No) so we can improve your experience in the future.

**Pull request closed without merging**